src/main/proto/netflix/titus/titus_cluster_membership_api.proto

```diff
@@ -0,0 +1,205 @@
+// Titus supervisor data model.
+
+message ChangeRecord {
+
+    string instanceId = 1;
+
+    // (Required) Detailed code for being in a specific state.
+    string reasonCode = 2;
+
+    // (Optional) Textual description accompanying the 'reasonCode'.
+    string reasonMessage = 3;
+
+    // Time when a transition to a state happened.
+    uint64 timestamp = 4;
+}
+
+message ServicePort {
+    uint32 portNumber = 1;
+
+    string protocol = 2;
+
+    bool secure = 3;
+
+    string description = 4;
+}
+
+// A descriptor of a single service instance.
+message ServiceInstance {
+
+    enum ServiceState {
+
+        // Service instance is starting, and its operational state is not known yet.
+        Starting = 0;
+
+        // Service is inactive, and does not participate in the leader election process. Predefined
+        // reason codes in this state:
+        // * 'outOfService' - the instance is healthy, but taken of out of service by an administrator
+        // * 'unhealthy' - the instance is not healthy
+        Inactive = 1;
+
+        // Service is fully activated, and if does not participate in the leader election process, is ready to
+        // take traffic.
+        Active = 2;
+    }
+
+    enum LeadershipState {
+        // Service is healthy, and participates in the leader election process, but is not the leader yet.
+        NonLeader = 0;
+
+        // Service became a leader, and is activating all components. It does not accept client requests yet.
+        LeaderActivating = 1;
+
+        // Service is a fully activated leader.
+        LeaderActivated = 2;
+    }
+
+    string instanceId = 1;
+
+    string instanceGroupId = 2;
+
+    // (Required) Service instance state
+    ServiceState state = 3;
+
+    // (Required) For services with a leader, indicates the leadership status of this instance. Otherwise, always
+    // set to `NonLeader`.
+    LeadershipState leadershipState = 4;
+
+    string ipAddress = 5;
+
+    repeated ServicePort serverPorts = 6;
+
+    // Collection of key/value pairs. Read-only special name spaces:
+    // * 'aws.deployment.XXX' - deployment data (ASG creation time, EC2 instance launch time, etc)
+    // * 'aws.asgTags.XXX' - service's ASG tags
+    // * 'aws.ec2Tags.XXX' - service's EC2 tags
+    // * 'release.XXX' - release data (number, build time, bake time)
+    map<string, string> labels = 7;
+}
+
+message ClusterMembershipEvent {
+    message ServiceInstanceUpdate {
+        ChangeRecord change = 1;
+        ServiceInstance previous = 2;
+        ServiceInstance after = 3;
+    }
+
+    message ServiceInstanceRemoved {
+        ChangeRecord change = 1;
+        ServiceInstance last = 2;
+    }
+
+    oneof Event {
+        ServiceInstanceUpdate serviceUpdated = 1;
+        ServiceInstanceRemoved serviceRemoved = 2;
+    }
+}
+
+// ----------------------------------------------------------------------------
+// GRPC services
+
+message Id {
+    string id = 1;
```

Comment:
Is this meant to be a generic ID or an instanceId? If the latter, wondering if it should be renamed.

Reply:
Yes, in the new control plane API spec we defined common data models, to avoid duplication. One of them was `Id`. It should be removed from this IDL before we release the final version.

Reply:
We cannot use common `Id` message due to conflicts. I have renamed this to `MemberId`.